src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -237,6 +218,112 @@ ShouldSyncTableMetadata(Oid relationId)
 }
 
 
+/*
+ * SyncMetadataSnapshotToNode does the following:
+ *  1. Sets the localGroupId on the worker so the worker knows which tuple in
+ *     pg_dist_node represents itself.
+ *  2. Recreates the distributed metadata on the given worker.
+ * If raiseOnError is true, it errors out if synchronization fails.
+ */
+static bool
+SyncMetadataSnapshotToNode(WorkerNode *workerNode, bool raiseOnError)
+{
+	char *extensionOwner = CitusExtensionOwnerName();
+
+	/* generate and add the local group id's update query */
+	char *localGroupIdUpdateCommand = LocalGroupIdUpdateCommand(workerNode->groupId);
+
+	/* generate the queries which drop the metadata */
+	List *dropMetadataCommandList = MetadataDropCommands();
+
+	/* generate the queries which create the metadata from scratch */
+	List *createMetadataCommandList = MetadataCreateCommands();
+
+	List *recreateMetadataSnapshotCommandList = list_make1(localGroupIdUpdateCommand);
+	recreateMetadataSnapshotCommandList = list_concat(recreateMetadataSnapshotCommandList,
+													  dropMetadataCommandList);
+	recreateMetadataSnapshotCommandList = list_concat(recreateMetadataSnapshotCommandList,
+													  createMetadataCommandList);
+
+	/*
+	 * Send the snapshot recreation commands in a single remote transaction and
+	 * if requested, error out in any kind of failure. Note that it is not
+	 * required to send createMetadataSnapshotCommandList in the same transaction
+	 * that we send nodeDeleteCommand and nodeInsertCommand commands below.
+	 */
+	if (raiseOnError)
+	{
+		SendCommandListToWorkerInSingleTransaction(workerNode->workerName,
+												   workerNode->workerPort,
+												   extensionOwner,
+												   recreateMetadataSnapshotCommandList);
+		return true;
+	}
+	else
+	{
+		bool success =
+			SendOptionalCommandListToWorkerInTransaction(workerNode->workerName,
+														 workerNode->workerPort,
+														 extensionOwner,
+														 recreateMetadataSnapshotCommandList);
+		return success;
+	}
+}
+
+
+/*
+ * SendOptionalCommandListToWorkerInTransaction sends the given command list to
+ * the given worker in a single transaction. If any of the commands fail, it
+ * rollbacks the transaction, and otherwise commits.
+ */
+bool
+SendOptionalCommandListToWorkerInTransaction(char *nodeName, int32 nodePort,
+											 char *nodeUser, List *commandList)
+{
+	MultiConnection *workerConnection = NULL;
+	ListCell *commandCell = NULL;
+	int connectionFlags = FORCE_NEW_CONNECTION;
+	bool failed = false;
+
+	if (XactModificationLevel > XACT_MODIFICATION_NONE)
```

Comment:
If this check is still relevant maybe move it to the caller such that this becomes a generic utility function.

Reply:
I think it can be removed, since `start_metadata_sync_to_node` cannot be run in a transaction.

Reply:
done